### PR TITLE
Batch deflation benchmark

### PR DESCRIPTION
**Pull request by JPRichings**
program takes in {Ls, redblack, minSourceBatchsize, maxSourceBatchsize, minEVBatchsize, maxEVBatchsize}

Scans all combinations and measures the speed of the memory of host and device.

### CODE REVIEW

**Review by aportelli — CHANGES_REQUESTED** [review]
Thanks! The function `projAccumulate` is duplicated, which is not ideal we want to test the one used in the library. Could you please change the benchmark so that it uses the real class?

**Comment by JPRichings** [comment]
Changes should now be ready for review.

File: Hadrons/Modules/MGuesser/BatchExactDeflation.hpp [hunk 1]
```diff
@@ -127,12 +179,19 @@ class BatchExactDeflationGuesser: public LinearFunction<typename Pack::Field>
...
-                 TensorRemove(innerProduct(epack_.evec[i], in[j]))/epack_.eval[i], 
-                 epack_.evec[i], out[j]);
+                 TensorRemove(innerProduct(epack.evec[i], in[j]))/epack.eval[i], 
+                 epack.evec[i], out[j]);
         }
         t += usecond();
         // performance (STREAM convention): innerProduct 2 reads + axpy 2 reads 1 write = 5 transfers
         LOG(Message) << "projAccumulate: " << t << " us | " << 5.*nIt*lSizeGB << " GB | " << 5.*nIt*lSizeGB/t*1.0e6 << " GB/s" << std::endl;
+    }
+private:
```
Comment (aportelli) [inline Hadrons/Modules/MGuesser/BatchExactDeflation.hpp:189]: This is not useful, please remove and use the static function in the main code

File: tests/Benchmark_BatchGuesser.cpp [hunk 2]
```diff
@@ -0,0 +1,269 @@
+#include <Grid/Grid.h>
+
+using namespace Grid;
+
+template<typename Field> 
+void projAccumulate(const std::vector<Field> &in, std::vector<Field> &out,
```
Comment (aportelli) [inline tests/Benchmark_BatchGuesser.cpp:6]: This should be replaced by the guesser function

File: tests/Benchmark_BatchGuesser.cpp [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+        eigenVec.resize(evBlockSize,g);
+        eigenval.resize(evBlockSize);
+
+        for (auto &e: eigenVec)
+        {
+            random(RNG5,e);
+        }
+
+        //for (unsigned int e = 0; e < eigenval.size(); e++)
```
Comment (aportelli) [inline tests/Benchmark_BatchGuesser.cpp:70]: remove dead commented code

File: tests/Benchmark_BatchGuesser.cpp [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        std::cout << GridLogMessage << "evBlockSize: " << evBlockSize << std::endl;
+
+        for(unsigned int j = 0; j < in.size(); j += sb)
+        {
+            unsigned int srcBlockSize = std::min((int)in.size() - j, sb);
+
+            std::cout << GridLogMessage << "srcBlockSize: " << srcBlockSize << std::endl;
+
+            w1.Start();
+            projAccumulate<Field>(in, out, eigenVec, eigenval, evBlockSize, j, j + srcBlockSize);
```
Comment (aportelli) [inline tests/Benchmark_BatchGuesser.cpp:87]: This should call the function in the guesser

File: tests/Benchmark_BatchGuesser.cpp [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+            w1.Reset();
+        }
+    }
+
+    std::cout << GridLogMessage << "ProjAccumRunner end" << std::endl;
+    std::cout << GridLogMessage << "ProjAccum total: " << ProjAccum << std::endl;
+
+}
+
+inline void makeGrid(std::shared_ptr<GridBase> &gPt,
```
Comment (aportelli) [inline tests/Benchmark_BatchGuesser.cpp:99]: This likely duplicates code, use the environment

File: Hadrons/Modules/MGuesser/BatchExactDeflation.hpp [hunk 6]
```diff
@@ -83,36 +84,87 @@ class BatchExactDeflationGuesser: public LinearFunction<typename Pack::Field>
         unsigned int sourceSize = out.size();
 
         LOG(Message) << "=== BATCH DEFLATION GUESSER START" << std::endl;
+        LOG(Message) << "Total Sources: " << sourceSize << std::endl;
```
Comment (aportelli) [inline Hadrons/Modules/MGuesser/BatchExactDeflation.hpp:87]: The timer logic here is quite messy and could be done in a much smaller amount of code I think. Hadrons has a TimerArray class exactly to streamline that kind of timing.

**Review by aportelli — CHANGES_REQUESTED** [review]
Thanks, it is not completely ready yet. The most puzzling thing is that we changed the function in the guesser to call it directly in the benchmark, but you are actually not doing it?
See other comment in the code, plus two more general comments
- [ ]: You use a lot of line breaks and it is not always consistent, please clean the code a bit
- [ ]: Don't commit commented code unless there is a strong justification for it
- [ ]: Use the Hadrons logger and not the Grid one

File: Hadrons/Modules/MGuesser/BatchExactDeflation.hpp (hunk 1 above)
Comment (JPRichings) [reply Hadrons/Modules/MGuesser/BatchExactDeflation.hpp]: removed

File: tests/Benchmark_BatchGuesser.cpp (hunk 2 above)
Comment (JPRichings) [reply tests/Benchmark_BatchGuesser.cpp]: now uses a static class member from the batch guesser module

File: tests/Benchmark_BatchGuesser.cpp (hunk 3 above)
Comment (JPRichings) [reply tests/Benchmark_BatchGuesser.cpp]: done

File: tests/Benchmark_BatchGuesser.cpp (hunk 5 above)
Comment (JPRichings) [reply tests/Benchmark_BatchGuesser.cpp]: not duplicate code but changed to use the hadrons environment.

File: Hadrons/Modules/MGuesser/BatchExactDeflation.hpp (hunk 6 above)
Comment (JPRichings) [reply Hadrons/Modules/MGuesser/BatchExactDeflation.hpp:88]: timers changed to Hadrons ones as recommended.

File: tests/Benchmark_BatchGuesser.cpp (hunk 4 above)
Comment (JPRichings) [reply tests/Benchmark_BatchGuesser.cpp]: old version of the benchmark see Hadrons/utilities

**Review by aportelli — APPROVED** [review]
Many thanks!

**Pull request merged**